### ISSUE TITLE
components using useContextSelector have to be wrapped in memo to prevent surplus re-renders

### ISSUE THREAD

**Issue by laurencefass** [dai-shi/use-context-selector#65]
components using use-context-selector have to be wrapped in memo to prevent surplus re-renders

demo here:

https://codesandbox.io/s/react-use-context-selector-mn8sv?file=/src/App.js

Is this by design, or is it a bug? I cant see any reason why the updates should occur in the non-memoized component.

This is not the case with Redux as components using useSelector do not have to be wrapped by default.

e.g. https://codesandbox.io/s/rtk-counter-53y5q

Thanks

**Comment by dai-shi** [comment]
It's how React works. When a component re-renders, all children re-render.
This is the pattern to avoid `memo`.
https://codesandbox.io/s/react-use-context-selector-forked-hp59j?file=/src/App.js
(It's the pattern not only for `use-context-selector` but for context in general.)

**Comment by laurencefass** [comment]
Thanks for feedback.

**Comment by laurencefass** [comment]
May I ask: why do you not conceal the useCallback hook call inside your useContextSelector function so its not a requirement on the consumer? It doesnt work properly without it. 

i.e. change from

```  
const counter1 = useContextSelector(
    context,
    useCallback((state) => state.state.counter1, [])
  );
```
to
```
// the wrapped function exposed to the package consumer
function useContextSelector_wrapped(context, callback) {
  return useContextSelector_existing(context,  useCallback(callback, [])
}
```
to conceal the additional useHook, and then use it as follows to match Redux pattern
```
const counter1 = useContextSelector_wrapped(context, (state) => state.state.counter1));
```

I guess my question is why do you expose the additional useHook call to the function consumer?

thanks

**Comment by dai-shi** [comment]
Oops, sorry, you don't need stable selectors for this lib.
You can just use inline selectors (= no useCallback).

**Comment by laurencefass** [comment]
Apologies but I dont understand that statement. What is inline selector and how do i use it can you include example? 
Thanks

**Comment by dai-shi** [comment]
Sorry, it just means you can do it as you did.
https://codesandbox.io/s/react-use-context-selector-forked-co6gy?file=/src/App.js

**Comment by dai-shi** [comment]
The OP sandbox is
```jsx
const App = () => {
  const [state, setState] = useState(...)
  return (
    <Context.Provider ...>
      ...
    </Context.Provider>
  )
}
```

When App re-renders all children re-render.
Again, it has nothing to do with how use-context-selector is implemented.
This is a tricky part of how React works/optimizes re-renders.
Maybe this article helps? https://overreacted.io/before-you-memo/

**Comment by laurencefass** [comment]
Thanks for your feedback and time.

**Comment by laurencefass** [comment]
I have realised that I mixed up sandbox links between two very similar packages and I see you actually helped me fix a similar issue with  react-use-context-selector which now works. And thankyou.

However the link I was meant to post was this one: https://codesandbox.io/s/use-context-selector-xbwvl, which doesnt yet work. Notice that the counter1 subscriber updates when counter2 is incremented. I tried the trick with the callback wrapper and that didnt work...

I see you dont appear to include a <Provider> wrapper (as does react-use-context-selector) and I am having the same problem as the previous. i.e. the unsubsribed components are re-rendering. I would be very grateful for any assistance and advice on fixing the above sandbox for use-context-selector.

Thanks!
